jina/peapods/runtimes/prefetch/base.py

```diff
@@ -1,76 +1,66 @@
 import argparse
 import asyncio
-from asyncio import Future
-from typing import AsyncGenerator, Dict, Union
+from typing import AsyncGenerator, List, Union, TYPE_CHECKING
 
 from ...grpc import Grpclet
-from ....helper import typename, get_or_reuse_loop
+from ....helper import typename
 from ....logging.logger import JinaLogger
 from ....types.message import Message
 
-__all__ = ['PrefetchCaller']
+__all__ = ['BasePrefetcher']
 
-if False:
+if TYPE_CHECKING:
     from ...zmq import AsyncZmqlet
+    from ....types.request import Request
+    from ....clients.base.http import HTTPClientlet
 
 
-class PrefetchCaller:
-    """An async zmq request sender to be used in the Gateway"""
+class BasePrefetcher:
+    """An base async request/response handler"""
 
     def __init__(
-        self, args: argparse.Namespace, iolet: Union['AsyncZmqlet', 'Grpclet']
+        self,
+        args: argparse.Namespace,
+        iolet: Union['AsyncZmqlet', 'Grpclet', 'HTTPClientlet'],
     ):
         """
         :param args: args from CLI
         :param iolet: One of AsyncZmqlet or Grpclet. Used for sending/receiving data to/from the Flow
         """
         self.args = args
-        self.name = args.name or self.__class__.__name__
-        self.logger = JinaLogger(self.name, **vars(args))
-        self._message_buffer: Dict[str, Future[Message]] = dict()
         self.iolet = iolet
+        self.receive_task = self._create_receive_task()
+        self.logger = JinaLogger(self.__class__.__name__, **vars(args))
 
-        if isinstance(iolet, Grpclet):
-            self.iolet.callback = self._unwrap_request
-            self._receive_task = get_or_reuse_loop().create_task(self.iolet.start())
-        else:
-            self._receive_task = get_or_reuse_loop().create_task(self._receive())
-
-    async def _unwrap_request(self, msg):
-        return self._process_message(msg.request)
-
-    async def _receive(self):
-        try:
-            while True:
-                message = await self.iolet.recv_message(callback=lambda x: x.response)
-                # during shutdown the socket will return None
-                if message is None:
-                    break
-
-                self._process_message(message)
-        except asyncio.CancelledError:
-            raise
-        finally:
-            for future in self._message_buffer.values():
-                future.cancel(
-                    'PrefetchCaller closed, all outstanding requests canceled'
-                )
-            self._message_buffer.clear()
-
-    def _process_message(self, message):
-        if message.request_id in self._message_buffer:
-            future = self._message_buffer.pop(message.request_id)
-            future.set_result(message)
-        else:
-            self.logger.warning(
-                f'Discarding unexpected message with request id {message.request_id}'
-            )
+    def _create_receive_task(self) -> asyncio.Task:
```

Comment:
add the `abc` abstract class so that it will never let u instantiate one without this being implemented

Reply:
done